<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Overview &mdash; GFFutils v0.5 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '#',
        VERSION:     '0.5',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="GFFutils v0.5 documentation" href="index.html" />
    <link rel="prev" title="Welcome to GFFutils’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to GFFutils’s documentation!"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">GFFutils v0.5 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>Documentation for this module can be found on the BitBucket wiki or in the
<tt class="docutils literal"><span class="pre">doc/build/html/index.html</span></tt> file.</p>
<p>This module is used for doing things with GFF files that are too
complicated for a simple <tt class="docutils literal"><span class="pre">awk</span></tt> or <tt class="docutils literal"><span class="pre">grep</span></tt> command line call.</p>
<p>For example, to get a BED file of genes from a GFF file, you can use something
simple like:</p>
<div class="highlight-python"><pre>grep "gene" chr2.gff | awk '{print $1,$4,$5}'</pre>
</div>
<p>But how would you use commandline tools to get a BED file of 3&#8217; exons from
genes longer than 5 kb?  Or how would you get the average number of isoforms
for genes on the plus strand?  These more complex questions are actually quite
easy to answer using <tt class="xref docutils literal"><span class="pre">GFFutils</span></tt> &#8211; see the <strong>Examples</strong> below for how it&#8217;s
done.</p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>Unzip the source code, and from the source directory run (with root
priveliges):</p>
<div class="highlight-python"><pre>python setup.py install</pre>
</div>
<p>Now you&#8217;re ready to create a GFF database and interact with it from a
Python shell like IPython.</p>
</div>
<div class="section" id="creating-a-gffdb">
<h1>Creating a GFFDB<a class="headerlink" href="#creating-a-gffdb" title="Permalink to this headline">¶</a></h1>
<p>For each GFF file you would like to use you need to create a GFF database.
This database is stored in a file, and is simply a sqlite3 database.  You
only have to do this once for each GFF file.  As long as you don&#8217;t delete
the database from your hard drive, you don&#8217;t have to do this time-consuming
step again.</p>
<p>The database will take roughly twice as much hard drive space as the
original text file.  This is the cost of interactivity and fast lookups.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">GFFutils</span>

<span class="c"># downloaded from, e.g., FlyBase</span>
<span class="n">gff_filename</span> <span class="o">=</span> <span class="s">&#39;/data/annotations/dm3.gff&#39;</span>

<span class="c"># the database that will be created</span>
<span class="n">db_filename</span> <span class="o">=</span> <span class="s">&#39;/data/dm3.db&#39;</span>

<span class="c"># do it! (this will take ~2 min to run)</span>
<span class="n">GFFutils</span><span class="o">.</span><span class="n">create_gffdb</span><span class="p">(</span><span class="n">gfffn</span><span class="p">,</span> <span class="n">db_filename</span><span class="p">)</span>
</pre></div>
</div>
<p>Now <tt class="docutils literal"><span class="pre">dm3.db</span></tt> is the sqlite3 database that can be used in all sorts of
weird and wonderful ways, outlined below.</p>
</div>
<div class="section" id="using-the-database-interactively">
<h1>Using the database interactively<a class="headerlink" href="#using-the-database-interactively" title="Permalink to this headline">¶</a></h1>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">GFFutils</span>

<span class="c"># Set up a GFFDB object, telling it the filename of your database as</span>
<span class="c"># created above</span>
<span class="n">G</span> <span class="o">=</span> <span class="n">GFFutils</span><span class="o">.</span><span class="n">GFFDB</span><span class="p">(</span><span class="s">&#39;dm3.db&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>For performance, most of the <tt class="xref docutils literal"><span class="pre">GFFDB</span></tt> class methods return iterators.
In practice, you will need to either convert them to a list or iterate
through them in a list comprehension or a for-loop.  You can also grab the
next item in an iterator with its <tt class="docutils literal"><span class="pre">.next()</span></tt> method.  All four ways of
getting info from an iterator are shown below in the examples.</p>
<div class="section" id="identifying-what-s-in-the-database">
<h2>Identifying what&#8217;s in the database<a class="headerlink" href="#identifying-what-s-in-the-database" title="Permalink to this headline">¶</a></h2>
<p>What sorts of features are in the db?  The <tt class="xref docutils literal"><span class="pre">GFFDB.features()</span></tt> method
returns an iterator of the featuretypes that were in the GFF file (and
which are now in the <tt class="docutils literal"><span class="pre">featuretype</span></tt> field of the sqlite3 database, which
this method accesses).</p>
<p>Most methods in a <tt class="xref docutils literal"><span class="pre">GFFDB</span></tt> object return iterators for performance.</p>
<p>Since this is the first example of using the iterators returned by a
<tt class="xref docutils literal"><span class="pre">GFFDB</span></tt> object, here are a few different ways to get the results
from the iterator it returns.</p>
<p>Method 0: Convert iterator to a list:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">featuretype_iterator</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">features</span><span class="p">()</span>
<span class="n">featuretypes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">featuretype_iterator</span><span class="p">)</span>
</pre></div>
</div>
<p>Method 1: Use iterator in a for-loop (preferred):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">featuretype_iterator</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">features</span><span class="p">()</span>
<span class="k">for</span> <span class="n">featuretype</span> <span class="ow">in</span> <span class="n">featuretype_iterator</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">featuretype</span>
</pre></div>
</div>
<p>Method 2: Call <tt class="xref docutils literal"><span class="pre">next()</span></tt> incrementally on the iterator:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">featuretype_iterator</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">features</span><span class="p">()</span>
<span class="n">featuretype_1</span> <span class="o">=</span> <span class="n">featuretype_iterator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="n">featuretype_2</span> <span class="o">=</span> <span class="n">featuretype_iterator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="n">featuretype_3</span> <span class="o">=</span> <span class="n">featuretype_iterator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="n">featuretype_4</span> <span class="o">=</span> <span class="n">featuretype_iterator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="o">...</span>
<span class="o">...</span>

<span class="n">featuretypes</span> <span class="o">=</span> <span class="p">[</span><span class="n">featuretype1</span><span class="p">,</span> <span class="n">featuretype2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<p>It&#8217;s mostly a matter of preference which method you use.  However, using
the for-loop approach is most memory-efficient, since only a single
featuretype is in memory at one time.  This is not too important for
iterating through featuretypes (of which there are usually &lt;50; typically
3-10).  But when you want to iterate through 15,000 genes it can be useful.</p>
<p>In any case, we get something like the following.  This depends entirely on
the GFF file that you created your database from:</p>
<div class="highlight-python"><pre>['BAC_cloned_genomic_insert',
 'CDS',
 'DNA_motif',
 'breakpoint',
 'chromosome_arm',
 'chromosome_band',
 'complex_substitution',
 'deletion',
 'enhancer',
 'exon',
 'five_prime_UTR',
 'gene',
 'insertion_site',
 'intron',
 ...
 ...
  'tRNA',
 'tandem_repeat',
 'three_prime_UTR',
 'transposable_element',
 'transposable_element_insertion_site',
 'uncharacterized_change_in_nucleotide_sequence']</pre>
</div>
</div>
<div class="section" id="retrieving-specific-feature-types">
<h2>Retrieving specific feature types<a class="headerlink" href="#retrieving-specific-feature-types" title="Permalink to this headline">¶</a></h2>
<p>To retrieve just genes, just exons, or any other feature type that was in
the GFF file, use the <tt class="xref docutils literal"><span class="pre">GFFDB.features_of_type()</span></tt> method.  This will return
an iterator of <tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> objects.  These objects are described in
more detail in another section below.</p>
<p><tt class="docutils literal"><span class="pre">'gene'</span></tt> was in the list of <tt class="docutils literal"><span class="pre">featuretypes</span></tt> above.  Let&#8217;s
find out how many genes there were:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gene_iterator</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">)</span>

<span class="c"># convert iterator to list so we can get a length</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gene_iterator</span><span class="p">)</span>

<span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span>
</pre></div>
</div>
<p>Here&#8217;s a more memory-efficient way to do the same thing.  In this method,
we&#8217;re not bringing ALL the genes into a giant list &#8211; we&#8217;ll just increment
a counter.  Only a single <tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> object is in memory at a
time, which is the advantage of iterators . . .</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gene_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">gene_count</span> <span class="o">+=</span> <span class="mf">1</span>
<span class="k">print</span> <span class="n">gene_count</span>
</pre></div>
</div>
<p>Feature types not found in the db will not return an error (maybe
they should, eventually?); they just don&#8217;t return anything:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ncabbages</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;cabbage&#39;</span><span class="p">)))</span>
<span class="k">print</span> <span class="n">ncabbages</span>  <span class="c"># zero cabbages.</span>
</pre></div>
</div>
<p>Already know the ID of a feature?  Get the <tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> object
for that gene directly like this:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">my_favorite_feature</span> <span class="o">=</span> <span class="n">G</span><span class="p">[</span><span class="s">&#39;FBgn0002121&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="gfffeatures-in-more-detail">
<h1>GFFFeatures in more detail<a class="headerlink" href="#gfffeatures-in-more-detail" title="Permalink to this headline">¶</a></h1>
<p>Just to make sure we&#8217;re on the same page, here&#8217;s the setup for this
section:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">GFFutils</span>
<span class="n">G</span> <span class="o">=</span> <span class="n">GFFutils</span><span class="o">.</span><span class="n">GFFDB</span><span class="p">(</span><span class="s">&#39;dm3.db&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Let&#8217;s get a single <tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> to work with.  Since I don&#8217;t know
any accessions off the top of my head, let&#8217;s just get the first gene in the
iterator:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">genes_iterator</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">)</span>
<span class="n">gene</span> <span class="o">=</span> <span class="n">genes_iterator</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
</pre></div>
</div>
<p><tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> objects, when printed, show useful information:</p>
<div class="highlight-python"><pre>GFFFeature gene 'FBgn0031208': chr2L:7529-9484 (+)
#           ^          ^              ^         ^
#           |          |              |         |
# featuretype      accession   genomic coords   strand</pre>
</div>
<p><tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> objects have an attribute, <tt class="docutils literal"><span class="pre">id</span></tt>, which contains the
accession in the attributes field of the original GFF file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span>

<span class="s">&#39;FBgn0031208&#39;</span>
</pre></div>
</div>
<p>They also have many other properties:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">start</span>
<span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">stop</span>
<span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">chr</span>
<span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">featuretype</span>
<span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">strand</span>
</pre></div>
</div>
<p>You can get the length of a gene with:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gene_len</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">stop</span> <span class="o">-</span> <span class="n">gene</span><span class="o">.</span><span class="n">start</span>
</pre></div>
</div>
<p>or you can use the perhaps-more-convenient:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gene_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</pre></div>
</div>
<p>In a <tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> object, the <tt class="xref docutils literal"><span class="pre">GFFFeature.attributes</span></tt>
attribute holds all the info that was in the attributes column of your GFF
file.  This will vary based on what was in your original GFF file.  You can
get a list of this with:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">_attrs</span>
</pre></div>
</div>
<p>and you can access any of the attributes with a dot, then the
attribute name.  For example, in the GFF file I used, the above code
returned:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="s">&#39;ID&#39;</span><span class="p">,</span> <span class="s">&#39;Name&#39;</span><span class="p">,</span> <span class="s">&#39;Ontology_term&#39;</span><span class="p">,</span> <span class="s">&#39;Dbxref&#39;</span><span class="p">,</span> <span class="s">&#39;derived_computed_cyto&#39;</span><span class="p">,</span> <span class="s">&#39;gbunit&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>So we could get the ontology terms for this gene with:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">Ontology_term</span>
</pre></div>
</div>
<p>Or the DBxref for the gene with:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">gene</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">Dbxref</span>
</pre></div>
</div>
<p>You can parse this info out yourself; parsing these into sub-attributes
of a <tt class="xref docutils literal"><span class="pre">GFFFeature.Attribute</span></tt> object is something I haven&#8217;t implemented
yet...</p>
<p>You now know enough to be able to generate a line for a BED-format
file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">line</span> <span class="o">=</span> <span class="s">&#39;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">chr</span><span class="p">,</span>
                                     <span class="n">gene</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                                     <span class="n">gene</span><span class="o">.</span><span class="n">stop</span><span class="p">,</span>
                                     <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                                     <span class="n">gene</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                                     <span class="n">gene</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="k">print</span> <span class="n">line</span>
</pre></div>
</div>
<p>But <tt class="xref docutils literal"><span class="pre">GFFFeature</span></tt> objects have a convenience function,
<tt class="xref docutils literal"><span class="pre">to_bed()</span></tt>, which also accepts a number from 3 to 6 so you can tell it
how many BED fields you want returned (3 fields is the default).</p>
<p>So you could write a BED file of all the genes like so:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">fout</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;genes.bed&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>  <span class="c"># open a file for writing</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">to_bed</span><span class="p">())</span>
<span class="n">fout</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>This can be extremely useful for downstream processing by, for example,
BEDtools.</p>
</div>
<div class="section" id="navigating-the-hierarchy-of-features">
<h1>Navigating the hierarchy of features<a class="headerlink" href="#navigating-the-hierarchy-of-features" title="Permalink to this headline">¶</a></h1>
<p>Here&#8217;s how to find the transcripts belonging to a gene.  The
<tt class="xref docutils literal"><span class="pre">GFFFeature.children()</span></tt> and <tt class="xref docutils literal"><span class="pre">GFFFeature.parents()</span></tt> methods need a
feature ID as an argument, which is stored in the <tt class="xref docutils literal"><span class="pre">GFFFeature.id</span></tt>
attribute:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
    <span class="k">print</span> <span class="n">i</span>
</pre></div>
</div>
<p>Here&#8217;s how to find the exons belonging to a gene.  By default, level=1,
which means a &#8216;hierarchy distance&#8217; of 1 (direct parent/children).  level=2
is analagous to grandparent/grandchild, which is used for the relationship
between genes/exons.  level=3 not currently implemented (not clear where it
would be used):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mf">2</span><span class="p">):</span>
    <span class="k">print</span> <span class="n">i</span>
</pre></div>
</div>
<p>Note that, depending on your GFF file, you may have more than just exons as
the children of genes (e.g., 3&#8217; UTRs, introns, 5&#8217; UTRs).  If you just want
the exons, then you can filter by feature type:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mf">2</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">featuretype</span> <span class="o">==</span> <span class="s">&#39;exon&#39;</span><span class="p">:</span>
        <span class="k">print</span> <span class="n">i</span>
</pre></div>
</div>
</div>
<div class="section" id="file-format-conversions">
<h1>File format conversions<a class="headerlink" href="#file-format-conversions" title="Permalink to this headline">¶</a></h1>
<p>Converting features to BED files was described above; briefly:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">fout</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;genes.bed&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">to_bed</span><span class="p">())</span>
<span class="n">fout</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Exporting a refFlat entry for one gene:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">G</span><span class="o">.</span><span class="n">refFlat</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
</pre></div>
</div>
<p>Create a new file, writing a refFlat entry for each gene.  Note that the
<tt class="xref docutils literal"><span class="pre">refFlat()</span></tt> method is set up such that it will return <tt class="xref docutils literal"><span class="pre">None</span></tt> if there
were no CDSs for a particular gene.  We don&#8217;t want to write these to file,
but do want to keep track of them.</p>
<p>This will take a few seconds to run:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">missing_cds</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">fout</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;mydatabase.refFlat&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">rflt</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">refFlat</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rflt</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">rflt</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">missing_cds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

<span class="n">fout</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>So, what were those genes that didn&#8217;t have CDSs?  Check the first 25:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">missing_cds</span><span class="p">[:</span><span class="mf">25</span><span class="p">]:</span>
    <span class="k">print</span> <span class="n">g</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">Name</span><span class="p">[</span><span class="mf">0</span><span class="p">]</span>
</pre></div>
</div>
<p>Ahhhhh . . . a bunch of snoRNAs, tRNAs, etc.  Makes sense!</p>
<p><tt class="xref docutils literal"><span class="pre">GFFFeatures</span></tt> have a <tt class="xref docutils literal"><span class="pre">GFFFeature.tostring()</span></tt> method which prints
back the GFF file entry as a string (with the newline included).  This
makes it very easy to write new GFF files containing a subset of the
features in the original GFF file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># new GFF file with genes &gt; 5kb</span>
<span class="n">fout</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;big-genes.gff&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">5000</span><span class="p">:</span>
        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">tostring</span><span class="p">())</span>
<span class="n">fout</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>In each case, assume the following setup:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">GFFutils</span>
<span class="n">GFFutils</span><span class="o">.</span><span class="n">create_gffdb</span><span class="p">(</span><span class="s">&#39;dm3.gff&#39;</span><span class="p">,</span><span class="s">&#39;dm3.db&#39;</span><span class="p">)</span>
<span class="n">G</span> <span class="o">=</span> <span class="n">GFFutils</span><span class="o">.</span><span class="n">GFFDB</span><span class="p">(</span><span class="s">&#39;dm3.db&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="gene-count">
<h2>Gene count<a class="headerlink" href="#gene-count" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gene_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;three_prime_UTR&#39;</span><span class="p">):</span>
    <span class="n">gene_count</span> <span class="o">+=</span> <span class="mf">1</span>
<span class="k">print</span> <span class="n">gene_count</span>
</pre></div>
</div>
</div>
<div class="section" id="average-gene-length">
<h2>Average gene length<a class="headerlink" href="#average-gene-length" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gene_lengths</span> <span class="o">=</span> <span class="mf">0</span>
<span class="n">gene_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">gene_lengths</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">gene_count</span> <span class="o">+=</span> <span class="mf">1</span>
<span class="n">mean_gene_length</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">gene_lengths</span><span class="p">)</span> <span class="o">/</span> <span class="n">gene_count</span>
<span class="k">print</span> <span class="n">mean_gene_length</span>
</pre></div>
</div>
</div>
<div class="section" id="longest-gene">
<h2>Longest gene<a class="headerlink" href="#longest-gene" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="n">maxlen</span> <span class="o">=</span> <span class="mf">0</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">gene_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gene_len</span> <span class="o">&gt;</span> <span class="n">maxlen</span><span class="p">:</span>
        <span class="n">maxlen</span> <span class="o">=</span> <span class="n">gene_len</span>
<span class="k">print</span> <span class="n">maxlen</span>
</pre></div>
</div>
</div>
<div class="section" id="average-exon-count">
<h2>Average exon count<a class="headerlink" href="#average-exon-count" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="n">exon_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="n">gene_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">gene_exon_count</span> <span class="o">=</span> <span class="mf">0</span>
    <span class="k">for</span> <span class="n">child</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="mf">2</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">child</span><span class="o">.</span><span class="n">featuretype</span> <span class="o">==</span> <span class="s">&#39;exon&#39;</span><span class="p">:</span>
            <span class="n">gene_exon_count</span> <span class="o">+=</span> <span class="mf">1</span>
    <span class="n">exon_count</span> <span class="o">+=</span> <span class="n">gene_exon_count</span>
    <span class="n">gene_count</span> <span class="o">+=</span> <span class="mf">1</span>
<span class="n">mean_exon_count</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">exon_count</span><span class="p">)</span> <span class="o">/</span> <span class="n">gene_count</span>
<span class="k">print</span> <span class="n">mean_exon_count</span>
</pre></div>
</div>
</div>
<div class="section" id="bed-file-of-3-exons-from-genes-longer-than-5-kb">
<h2>BED file of 3&#8217; exons from genes longer than 5 kb<a class="headerlink" href="#bed-file-of-3-exons-from-genes-longer-than-5-kb" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="n">fout</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;3prime-exons.bed&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;track name=&quot;3-prime exons</span><span class="se">\n</span><span class="s">&quot;&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">5000</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="n">children</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="mf">2</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">featuretype</span><span class="o">==</span><span class="s">&#39;exon&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s">&#39;+&#39;</span><span class="p">:</span>
        <span class="n">three_prime_exon</span> <span class="o">=</span> <span class="n">children</span><span class="p">[</span><span class="mf">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">three_prime_exon</span> <span class="o">=</span> <span class="n">children</span><span class="p">[</span><span class="o">-</span><span class="mf">1</span><span class="p">]</span>
    <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">three_prime_exon</span><span class="o">.</span><span class="n">to_bed</span><span class="p">())</span>
<span class="n">fout</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="histogram-of-exon-lengths">
<h2>Histogram of exon lengths<a class="headerlink" href="#histogram-of-exon-lengths" title="Permalink to this headline">¶</a></h2>
<p>(Assumes you have matplotlib installed)</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">p</span>
<span class="n">lengths</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">stop</span><span class="o">-</span><span class="n">i</span><span class="o">.</span><span class="n">start</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;exon&#39;</span><span class="p">)]</span>
<span class="n">p</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">lengths</span><span class="p">,</span><span class="n">bins</span><span class="o">=</span><span class="mf">50</span><span class="p">)</span>
<span class="n">p</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="average-number-of-isoforms-for-genes-on-plus-strand">
<h2>Average number of isoforms for genes on plus strand<a class="headerlink" href="#average-number-of-isoforms-for-genes-on-plus-strand" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre><span class="n">isoform_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="n">gene_count</span> <span class="o">=</span> <span class="mf">0</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s">&#39;-&#39;</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">featuretype</span><span class="o">==</span><span class="s">&#39;mRNA&#39;</span><span class="p">]</span>
    <span class="n">isoform_count</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">isoforms</span><span class="p">)</span>
    <span class="n">gene_count</span> <span class="o">+=</span> <span class="mf">1</span>
<span class="n">mean_isoform_count</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">isoform_count</span><span class="p">)</span> <span class="o">/</span> <span class="n">gene_count</span>
</pre></div>
</div>
</div>
<div class="section" id="constituitively-expressed-exons">
<h2>Constituitively expressed exons<a class="headerlink" href="#constituitively-expressed-exons" title="Permalink to this headline">¶</a></h2>
<p>The trick here is that if all of a gene&#8217;s isoforms are found in the list of
parents of an exon, then that exon is found in all isoforms of that gene.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">fout</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;constituitively-expressed-exons.gff&#39;</span><span class="p">,</span><span class="s">&#39;w&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">):</span>
    <span class="n">const_expr_exons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">isoforms</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">featuretype</span><span class="o">==</span><span class="s">&#39;mRNA&#39;</span><span class="p">]</span>
    <span class="n">isoform_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">isoforms</span><span class="p">]</span>
    <span class="n">children</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="n">level</span><span class="o">=</span><span class="mf">2</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">child</span> <span class="ow">in</span> <span class="n">children</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">child</span><span class="o">.</span><span class="n">featuretype</span> <span class="o">!=</span> <span class="s">&#39;exon&#39;</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">exon_transcript_parent_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">parents</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="p">,</span><span class="n">level</span><span class="o">=</span><span class="mf">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">featuretype</span><span class="o">==</span><span class="s">&#39;mRNA&#39;</span><span class="p">]</span>

        <span class="c"># check to make sure each of the gene&#39;s isoforms are in this exon.</span>
        <span class="n">in_all_isoforms</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">for</span> <span class="n">i_id</span> <span class="ow">in</span> <span class="n">isoform_ids</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">exon_transcript_parent_ids</span><span class="p">:</span>
                <span class="n">in_all_isoforms</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">if</span> <span class="n">in_all_isoforms</span><span class="p">:</span>
            <span class="n">const_expr_exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">child</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">const_expr_exons</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0</span><span class="p">:</span>
        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">tostring</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">isoform</span> <span class="ow">in</span> <span class="n">isoforms</span><span class="p">:</span>
            <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">isoform</span><span class="o">.</span><span class="n">tostring</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">const_expr_exons</span><span class="p">:</span>
            <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">tostring</span><span class="p">())</span>
<span class="n">fout</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <h3><a href="index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="#">Overview</a></li>
<li><a class="reference external" href="#installation">Installation</a></li>
<li><a class="reference external" href="#creating-a-gffdb">Creating a GFFDB</a></li>
<li><a class="reference external" href="#using-the-database-interactively">Using the database interactively</a><ul>
<li><a class="reference external" href="#identifying-what-s-in-the-database">Identifying what&#8217;s in the database</a></li>
<li><a class="reference external" href="#retrieving-specific-feature-types">Retrieving specific feature types</a></li>
</ul>
</li>
<li><a class="reference external" href="#gfffeatures-in-more-detail">GFFFeatures in more detail</a></li>
<li><a class="reference external" href="#navigating-the-hierarchy-of-features">Navigating the hierarchy of features</a></li>
<li><a class="reference external" href="#file-format-conversions">File format conversions</a></li>
<li><a class="reference external" href="#examples">Examples</a><ul>
<li><a class="reference external" href="#gene-count">Gene count</a></li>
<li><a class="reference external" href="#average-gene-length">Average gene length</a></li>
<li><a class="reference external" href="#longest-gene">Longest gene</a></li>
<li><a class="reference external" href="#average-exon-count">Average exon count</a></li>
<li><a class="reference external" href="#bed-file-of-3-exons-from-genes-longer-than-5-kb">BED file of 3&#8217; exons from genes longer than 5 kb</a></li>
<li><a class="reference external" href="#histogram-of-exon-lengths">Histogram of exon lengths</a></li>
<li><a class="reference external" href="#average-number-of-isoforms-for-genes-on-plus-strand">Average number of isoforms for genes on plus strand</a></li>
<li><a class="reference external" href="#constituitively-expressed-exons">Constituitively expressed exons</a></li>
</ul>
</li>
</ul>

            <h4>Previous topic</h4>
            <p class="topless"><a href="index.html"
                                  title="previous chapter">Welcome to GFFutils&#8217;s documentation!</a></p>
            <h3>This Page</h3>
            <ul class="this-page-menu">
              <li><a href="_sources/overview.txt"
                     rel="nofollow">Show Source</a></li>
            </ul>
          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to GFFutils’s documentation!"
             >previous</a> |</li>
        <li><a href="index.html">GFFutils v0.5 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
      &copy; Copyright 2010, Ryan Dale.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.5.
    </div>
  </body>
</html>